core/sile.lua

```diff
@@ -240,10 +241,251 @@ function SILE.resolveFile(filename, pathprefix)
   return nil
 end
 
+SILE.currentCommandStack = {
+  --- Internal: push-pop balance checking ID
+  _lastPushId = 0,
+  --- Internal: Function assigned to frames to convert them to string
+  _frameToString = function(self, skipFile, withAttrs)
+    local str
+    if skipFile or not self.file then
+      str = ""
+    else
+      str = self.file .. " "
+    end
+
+    if self.line then
+      str = str .. "l." .. self.line .. " "
+      if self.column then
+        str = str .. "col." .. self.column .. " "
+      end
+    end
+
+    if self.tag then
+      -- Command
+      str = str .. "\\" .. self.tag
+      if withAttrs then
+        str = str .. "["
+        local first = true
+        for key, value in pairs(self.options) do
+          if first then first = false else str = str .. ", " end
+          str = str .. key .. "=" .. value
+        end
+        str = str .. "]"
+      end
+    elseif self.text then
+      -- Literal string
+      local text = self.text
+      if text:len() > 20 then
+        text = text:sub(1, 18) .. "…"
+      end
+      text = text:gsub("\n", "␤"):gsub("\t", "␉"):gsub("\v", "␋")
+
+      str = str .. "\"" .. text .. "\""
+    else
+      -- Unknown
+      str = str .. type(self.content) .. "(" .. self.content .. ")"
+    end
+
+    --str = str .. "\n\n" .. self.pushTrace .. "\n"
+
+    return str
+  end
+}
+--- Push a command frame on to the stack to record the execution trace for debugging.
+--- Carries information about the command call, not the command itself.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushCommand(tag, line, column, options, file)
```

Comment:
Is there any time we push anything to the command stack that is not a command?

Reply:
Yes, text and possibly other stuff, after merging with breadcrumbs or if such need arises.